api/stock.js

```diff
@@ -1,56 +1,83 @@
 import fetch from 'node-fetch'
+import { fetchSymbol } from './_services/symbol.services'
 require('dotenv').config()
 
-const convertMsToS = n => n / 1000
+const convertMsToS = (n) => n / 1000
 
-const getTimestampFromDateStr = dt => convertMsToS(Math.floor(dt.getTime()))
+const getTimestampFromDateStr = (dt) => convertMsToS(Math.floor(dt.getTime()))
 
 export default async (request, response) => {
     const API_ENDPOINT = 'https://finnhub.io/api/v1/stock/candle'
+    const API_KEY = process.env.FINNHUB_API_KEY
+    const FIRST_DAY_2020 = '01/02/2020'
+    const CURRENT_DAY_MIDNIGHT = new Date().setHours(0, 0, 0, 0)
+
     const querySymbols = request.query.symbols
     const symbolList = querySymbols ? querySymbols.split(',') : []
 
     if (symbolList.length > 0) {
-        const API_KEY = process.env.FINNHUB_API_KEY
-        const FIRST_DAY_2020 = '01/02/2020'
-        const CURRENT_DAY_MIDNIGHT = new Date().setHours(0, 0, 0, 0)
+        const verifySymbolPromises = symbolList.map((symbol) =>
+            fetchSymbol(symbol)
+        )
 
-        const promises = symbolList.map(symbol => {
-            const from = getTimestampFromDateStr(new Date(FIRST_DAY_2020))
-            const to = convertMsToS(CURRENT_DAY_MIDNIGHT)
+        let verifySymbolResults = []
+        try {
+            verifySymbolResults = await Promise.all(verifySymbolPromises)
+        } catch (e) {
+            response.status(500)
+        }
 
-            return fetch(
-                `${API_ENDPOINT}?symbol=${symbol.toUpperCase()}&resolution=D&from=${from}&to=${to}&token=${API_KEY}`
-            ).then(res => res.json())
-        })
+        const fetchSymbolDataPromises = verifySymbolResults.map(
+            ({ symbol, found }) => {
+                if (found) {
+                    const from = getTimestampFromDateStr(
+                        new Date(FIRST_DAY_2020)
+                    )
+                    const to = convertMsToS(CURRENT_DAY_MIDNIGHT)
+
+                    return fetch(
+                        `${API_ENDPOINT}?symbol=${symbol.toUpperCase()}&resolution=D&from=${from}&to=${to}&token=${API_KEY}`
+                    ).then((res) => res.json())
+                }
+
+                return Promise.resolve(null)
+            }
+        )
 
-        let results = []
+        let fetchSymbolDataResults = []
 
         try {
-            results = await Promise.all(promises)
+            fetchSymbolDataResults = await Promise.all(fetchSymbolDataPromises)
         } catch (e) {
             response.status(500)
         }
 
         let data = []
         let dates = []
 
-        results.map((res, idx) => {
-            const { c, o, h, l, t, s } = res
+        fetchSymbolDataResults.map((res, idx) => {
             const symbol = symbolList[idx]
 
-            if (s !== 'no_data') {
+            if (res) {
+                const { c, o, h, l, t, s } = res
```

Comment:
Maybe we could populate the symbols of those not found when creating fetchSymbolDataPromises instead of just Promise.resolve(null). It would mean we could use this symbol instead of having to do symbolList[idx] which might not work all the time.

Reply:
Hmm not sure about this. We'll discuss